Core/include/Acts/Utilities/Any.hpp

```diff
@@ -0,0 +1,495 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2021 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include <any>
+#include <array>
+#include <cassert>
+#include <cstddef>
+#include <memory>
+#include <utility>
+
+// #define _ACTS_ANY_ENABLE_VERBOSE
+// #define _ACTS_ANY_ENABLE_DEBUG
+// #define _ACTS_ANY_ENABLE_TRACK_ALLOCATIONS
+
+#if defined(_ACTS_ANY_ENABLE_TRACK_ALLOCATIONS)
+#include <iostream>
+#include <mutex>
+#include <set>
+#include <typeindex>
+#include <typeinfo>
+#endif
+
+#if defined(_ACTS_ANY_ENABLE_VERBOSE) || defined(_ACTS_ANY_ENABLE_DEBUG)
+#include <iomanip>
+#include <iostream>
+#endif
+
+#if defined(_ACTS_ANY_ENABLE_DEBUG)
+#define _ACTS_ANY_DEBUG(x) std::cout << x << std::endl;
+#else
+#define _ACTS_ANY_DEBUG(x)
+#endif
+
+#if defined(_ACTS_ANY_ENABLE_VERBOSE)
+#define _ACTS_ANY_VERBOSE(x) std::cout << x << std::endl;
+#define _ACTS_ANY_VERBOSE_BUFFER(s, b) \
+  do {                                 \
+    std::cout << "" << s << ": 0x";    \
+    for (char c : b) {                 \
+      std::cout << std::hex << (int)c; \
+    }                                  \
+    std::cout << std::endl;            \
+  } while (0)
+#else
+#define _ACTS_ANY_VERBOSE(x)
+#define _ACTS_ANY_VERBOSE_BUFFER(s, b)
+#endif
+
+namespace Acts {
+
+#if defined(_ACTS_ANY_ENABLE_TRACK_ALLOCATIONS)
+static std::mutex _s_any_mutex;
+static std::set<std::pair<std::type_index, void*>> _s_any_allocations;
+
+#define _ACTS_ANY_TRACK_ALLOCATION(T, heap)                                  \
+  do {                                                                       \
+    std::lock_guard guard{_s_any_mutex};                                     \
+    _s_any_allocations.emplace(std::type_index(typeid(T)), heap);            \
+    _ACTS_ANY_DEBUG("Allocate type: " << typeid(T).name() << " at " << heap) \
+  } while (0)
+
+struct _AnyAllocationReporter {
+  ~_AnyAllocationReporter() {
+    std::lock_guard guard{_s_any_mutex};
+
+    if (!_s_any_allocations.empty()) {
+      std::cout << "Not all allocations have been released" << std::endl;
+      for (const auto& [idx, addr] : _s_any_allocations) {
+        std::cout << "- " << idx.name() << ": " << addr << std::endl;
+      }
+      std::abort();
+    }
+  }
+};
+static _AnyAllocationReporter s_reporter;
```

Comment:
This destructor will not run if the application aborts due to an allocator failure, as is currently happening... so I'm not sure if this is the right solution, although it is certainly an elegant one.

Reply:
That's true, but I don't have a better idea how to add a end of job hook without using an exit handler. Do you?

Reply:
I would borrow an idea from the Linux RCU synchronization primitive and have "checkpoints" throughout your test code that are invoked at strategic points of the test code where you assert that no allocation should be live.

Reply:
Hm, but then it's not automatic, and the idea was that this is transparent and global.

Reply:
Note that you can keep the automatic report, it does no harm. But with a manual report, you get more targeted and abort-immune info.

Reply:
Looks good to me overall, why do you have both _ACTS_ANY_CHECK_ALLOCATIONS (in the Any header) and CHECK_ANY_ALLOCATIONS (in the tests) doing the same thing ?

Reply:
I just messed up 😅. Removes the ones from `Any.hpp` now.

Reply:
Alright, let's merge this :)